experimental/tzif/src/parse/posix.rs

```diff
@@ -0,0 +1,756 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{
+    core::{
+        ext::{
+            byte::ParseBytes,
+            chain::Chain,
+            default::OrDefaultTo,
+            inspect::Inspect,
+            many::ParseMany,
+            map::{MapInner, MapValue, TryMapValue},
+            maybe::MaybeParse,
+            require::Require,
+            then::{OnSuccess, Then},
+        },
+        ParseInput, ParseResult, Parsed,
+    },
+    data::{
+        posix::{DstTransitionInfo, PosixTzString, TransitionDate, TransitionDay, ZoneVariantInfo},
+        time::{Hours, Minutes, Seconds},
+    },
+};
+use eyre::Context;
+
+impl<T, Source> PosixTzStringParser<Source> for T
+where
+    Source: Clone + ParseInput<u8, Vec<u8>, Source>,
+    T: ParseInput<u8, Vec<u8>, Source>,
+{
+}
+
+/// A trait extension for parsing POSIX time-zone strings as specified by
+/// <https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html>
+pub trait PosixTzStringParser<Source: Clone + ParseInput<u8, Vec<u8>, Source>>:
+    ParseInput<u8, Vec<u8>, Source>
+{
+    /// The string specifies the name of the time zone variant. It must be three or more characters
+    /// long and must not contain a leading colon, embedded digits, commas, nor plus and minus signs.
+    /// There is no space character separating the time zone name from the offset, so these restrictions
+    /// are necessary to parse the specification correctly.
+    ///
+    /// An arbitrary name may be specified by surrounding it with angled brackets, e.g. `<name>`.
+    fn parse_zone_variant_name(&mut self) -> ParseResult<String, Source> {
+        self.inspect_next()
+            .then_parse_with_context(|byte, source| match byte {
+                b'<' => source.parse_arbitrary_zone_variant_name(),
+                _ => source.parse_alphabetic_zone_variant_name(),
+            })
+            .wrap_err("parse_zone_variant_name(): failed to parse zone variant name")
+    }
+
+    /// Parses an arbitrary time-zone variant name. This name must be enclosed in angled brackets, e.g. `<name>`.
+    fn parse_arbitrary_zone_variant_name(&mut self) -> ParseResult<String, Source> {
+        self.begin_chain()
+            .chain(|source| source.expect_byte(b'<'))
```

Comment:
issue: I'm not a huge fan of this tuple based parsing system: IME what most parser libraries do is sequentially parse things as separate statements and you don't need complex tuple impls to make that work